tuf/repository_tool.py

```diff
@@ -3170,18 +3149,26 @@ def load_repository(repository_directory, repository_name='default',
     roleinfo['paths'] = metadata_object['targets']
     roleinfo['delegations'] = metadata_object['delegations']
 
-    tuf.roledb.add_role(metadata_name, roleinfo, repository_name)
-    loaded_metadata.append(metadata_name)
+    tuf.roledb.add_role(rolename, roleinfo, repository_name)
+
+    # Generate the Targets object of the delegated role,
+    # add it to the top-level 'targets' object and to its
+    # direct delegating role object.
+    new_targets_object = Targets(targets_directory, rolename,
+         roleinfo, parent_targets_object=targets_objects['targets'],
+         repository_name=repository_name)
+
+    targets_objects[rolename] = new_targets_object
 
-    # Generate the Targets objects of the delegated roles of 'metadata_name'
-    # and add it to the top-level 'targets' object.
-    new_targets_object = Targets(targets_directory, metadata_name, roleinfo,
-        repository_name=repository_name)
-    targets_object = targets_objects['targets']
-    targets_objects[metadata_name] = new_targets_object
+    targets_objects['targets'].add_delegated_role(rolename,
+        new_targets_object)
+    targets_objects[delegating_role].add_delegated_role(rolename,
+        new_targets_object)
 
-    targets_object._delegated_roles[(os.path.basename(metadata_name))] = \
-        new_targets_object
+    # Append the next level delegations to the list:
+    # the 'delegated' role becomes the 'delegating'
+    for delegation in metadata_object['delegations']['roles']:
+      delegations.append([delegation['name'], rolename])
```

Comment:
Wow, this is cool stuff! But I admit it took me a while to understand what's happening. You're adding delegations to the list of delegations as you iterate over them, thus traversing the tree of delegations.  :)

Out of curiosity, is there a name for this pattern? 

At any rate, we should add a big comment that describes it above the outer loop definition (L3123). And we should definitely point out that cycles in the delegation graph, will make this an infinite loop.

Reply:
Let's call it a custom breadth-first search :grin: 
I tried to improve it after you pointed out the potential infinite loop. 
The loaded delegations were still kept in the list of delegations anyway so the options to avoid repetitions seem to be:
- `if x is in list` - O(n)
- replace the list with a dictionary for O(1) lookup but dictionaries are guaranteed to be ordered only in Python 3.x so the traversal won't work
- use a `deque` for traversing the delegations (keep only the next in line) + `set` for storing already loaded ones

If this seems unnecessary overhead since cycles in the delegations are not likely to happen, we can keep only the `deque` and get rid of the `set` + adding a comment.

Hopefully the new comments help to make things look less magical :slightly_smiling_face: 
(and this is the new commit c79ea55)

Reply:
Haha, I think it just became even more magical. 😆 

Regarding the options:

Using a set as an infinite loop safe-guard seems like a good idea. And definitely better than using a list.
I think we would want to warn the caller if we encounter a loop, what do you think? To me, loops in the delegation graph don't seem desirable. But I may be missing something.

The need for `deque` is not fully obvious to me. Are you doing this to save memory? If so, I suppose that's fine. Maybe you can leave a note about the motivation. :)

Reply:
> Using a set as an infinite loop safe-guard seems like a good idea. And definitely better than using a list.
> I think we would want to warn the caller if we encounter a loop, what do you think? To me, loops in the delegation graph don't seem desirable. But I may be missing something.

It could be better to warn during the creation of the delegation ... but I think currently there is no mechanism to detect that so ...  better late than never?